---
title: 'CLARITY: Documentation of data sets'
author: "N. Melzer, D. Wittenburg"
date: "September 29, 2025"
output:
  word_document: default
  pdf_document: default
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo =TRUE)
library(DT)
options(DT.options = list(pageLength = 100, language = list(search = 'Filter:'))) ## used to show all rows otherwise only the first ten rows are displayed
path.input <- "inst/extdata/Fleckvieh"
path.input2 <- "inst/extdata/general"
```

&nbsp;&nbsp;&nbsp;

# Input data format required for CLARITY app
&nbsp;

## 1. Data sets generated applying the [hsrecombi](https://github.com/wittenburg/hsrecombi) pipeline.

&nbsp;

### 1.1 geneticMap.Rdata

data.frame *geneticMap*

```{r table1, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
load(file.path(path.input, "geneticMap.Rdata"))

df=cbind(1:ncol(geneticMap),colnames(geneticMap))

Class=c()
for(i in 1:ncol(geneticMap))Class[i]=class(geneticMap[,i])
df=cbind(df,Class)
                   
 colnames(df)<-c("Column no.","Column name","Class")                  
 
# output the table in a format suitable for HTML/PDF/docx conversion
datatable(df,rownames=F,options=list(searching=F,dom=""))
```

<font size="2">
Name: SNP name <br>
Chr: Chromosome <br>
Mbp_position: Chromosome position in megabase pair <br>
bp_position: Chromosome position in base pair <br>
L_cM: Chromosome position in centiMorgan relying on a likelihood approach<br>
D_cM: Chromosome position in centiMorgan relying on a deterministic approach <br>
D_recrate_adjacent: Recombination rate adjacent markers relying on a deterministic approach <br>
Mbp_inter_marker_distance: Inter-marker distance in Mbp <br>
Hm_cM: Chromosome position in centiMorgan relying on an HMM-based approach (male) <br>
Hm_recrate_adjacent: Recombination rate adjacent markers relying on an HMM-based approach (male) <br>
Hf_cM: Chromosome position in centiMorgan relying on an HMM-based approach (female) <br>
Hf_recrate_adjacent: Recombination rate adjacent markers relying on an HMM-based approach (female) <br>
Ha_cM: Chromosome position in centiMorgan relying on an HMM-based approach (sex-average) <br>
Ha_recrate_adjacent: Recombination rate adjacent markers relying on an HMM-based approach (sex-average) <br>
</font> 
&nbsp;&nbsp;


### 1.2 genetic_map_summary.Rdata
data.frame *genetic_map_summary*

```{r table2, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
load(file.path(path.input, "genetic_map_summary.Rdata"))

df=cbind(1:ncol(genetic_map_summary),colnames(genetic_map_summary))

Class=c()
for(i in 1:ncol(genetic_map_summary))Class[i]=class(genetic_map_summary[,i])
df=cbind(df,Class)
              
 colnames(df)<-c("Column no.","Column name","Class")                  
 
# output the table in a format suitable for HTML/PDF/docx conversion
datatable(df,rownames=F,options=list(searching=F,dom=""))
```


<font size="2"> 
Chr: Chromosome <br>
nSNP: Number of SNPs <br>
max_bp: Chromosome length in base pairs <br>
Gap_bp: Maximum gap size between pairs of adjacent markers in bp <br>
Space_kb: Inter-marker space in kilobase (kb) <br>
D_nRec: Number of cross-overs detected relying on deterministic approach<br>
D_M: Genetic length in Morgan estimated based on deterministic approach <br>
D_cMMb: centiMorgan per megabase pair for the deterministic approach <br>
L_M: Genetic length in Morgan estimated with the likelihood-based approach <br>
L_cMMb: centiMorgan per megabase pair for the likelihood-based approach <br>
Hm_nRec: Number of cross-overs detected relying on an HMM-based approach (male)<br>
Hm_M: Genetic length in Morgan estimated relying on an HMM-based approach (male) <br>
Hm_cMMb: centiMorgan per megabase pair for an HMM-based approach (male) <br>
Hf_nRec: Number of cross-overs detected relying on an HMM-based approach (female)<br>
Hf_M: Genetic length in Morgan estimated relying on an HMM-based approach (female) <br>
Hf_cMMb: centiMorgan per megabase pair for an HMM-based approach (female) <br>
Ha_nRec: Number of cross-overs detected relying on an HMM-based approach (sex-average)<br>
Ha_M: Genetic length in Morgan estimated relying on an HMM-based approach (sex-average) <br>
Ha_cMMb: centiMorgan per megabase pair for an HMM-based approach (sex-average) <br>
</font> 

&nbsp;&nbsp;


### 1.3 adjacentRecRate.Rdata

data.frame *adjacentRecRate*

```{r table3, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
load(file.path(path.input, "adjacentRecRate.Rdata"))

df=cbind(1:ncol(adjacentRecRate),colnames(adjacentRecRate))

Class=c()
for(i in 1:ncol(adjacentRecRate))Class[i]=class(adjacentRecRate[,i])
df=cbind(df,Class)
                   
colnames(df)<-c("Column no.","Column name","Class")                  
 
# output the table in a format suitable for HTML/PDF/docx conversion
datatable(df,rownames=F,options=list(searching=F,dom=""))
```

<font size="2"> 
Chr: Chromosome <br>
SNP: SNP name <br>
BP: Chromosome position in base pairs <br>
D_cM:  Chromosome position in centiMorgan based on deterministic approach <br>
D_Theta: Recombination rate based on deterministic approach <br>
Hm_cM: Chromosome position in centiMorgan relying on an HMM-based approach (male)<br>
Hm_Theta: Recombination rate relying on an HMM-based approach (male)<br>
Hf_cM: Chromosome position in centiMorgan relying on an HMM-based approach (female)<br>
Hf_Theta: Recombination rate relying on an HMM-based approach (female)<br>
Ha_cM: Chromosome position in centiMorgan relying on an HMM-based approach (sex-average)<br>
Ha_Theta: Recombination rate relying on an HMM-based approach (sex-average)<br>
Dis: Distance to preceeding marker <br>
</font>

&nbsp;&nbsp;

### 1.4 bestmapfun.Rdata

matrix *out*

```{r table4, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
load(file.path(path.input, "bestmapfun.Rdata"))

df=cbind(1:ncol(out),colnames(out))

Class=c()
for(i in 1:ncol(out))Class[i]=class(out[,i])
df=cbind(df,Class)
                   
colnames(df)<-c("Column no.","Column name","Class")                  
 
# output the table in a format suitable for HTML/PDF/docx conversion
datatable(df,rownames=F,options=list(searching=F,dom=""))
```

<font size="2"> 
Chr: Chromosome <br>
mse: Mean squared error <br>
par: Parameter <br>
</font>

&nbsp;&nbsp;


### 1.5 curve-short-\<chr\>.Rdata

$\text{<chr>}=1,\ldots,29$; list *store* with four elements:

1. matrix: dist_M and theta
2. matrix: x-values for Haldane scaled, Rao, Felsenstein, and Liberman & Karlin
3. matrix: y-values  for Haldane scaled, Rao, Felsenstein, and Liberman & Karlin
4. numeric value: the percentage of marker pairs involved in scatter plot of dist_M and theta


&nbsp;&nbsp;

## 2. Additional generated data sets

### 2.1  OverviewBreeds.Rdata

data.frame *BreedOverview*

```{r table5, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
load(file.path(path.input2, "OverviewBreeds.Rdata"))
out=OverviewBreed
df=cbind(1:ncol(out),colnames(out))

Class=c()
for(i in 1:ncol(out))Class[i]=class(out[,i])
df=cbind(df,Class)
                   
colnames(df)<-c("Column no.","Column name","Class")                  
 
# output the table in a format suitable for HTML/PDF/docx conversion
datatable(df,rownames=F,options=list(searching=F,dom=""))
```

<font size="2"> 
n1: Number of genotyped animals <br>
N1: Number of half-sib families <br>
n2: Number of genotyped animals in half-sib families with at least 30 progeny <br>
N2: Half-sib families with sires having at least 30 progenies <br>
p: Number of SNPs <br>
</font> 

&nbsp;&nbsp;

### 2.2 misplaced_all_breeds.Rdata

tibble *tab*

```{r table7, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
load(file.path(path.input2, "misplaced_all_breeds.Rdata"))
out=data.frame(tab)
df=cbind(1:ncol(out),colnames(out))

Class=c()
for(i in 1:ncol(out))Class[i]=class(out[,i])
df=cbind(df,Class)
                   
colnames(df)<-c("Column no.","Column name","Class")                  
 
# output the table in a format suitable for HTML/PDF/docx conversion
datatable(df,rownames=F,options=list(searching=T,dom=""))

```
<font size="2"> 
leg.breeds.i: Breed name <br>
Chr: Chromosome <br>
Name: SNP name <br>
Mbp_position: Chromosomal position in million base pairs <br>
</font> 

&nbsp;&nbsp;

### 2.3 misplacedMarkers.Rdata

tibble *misplacedMarkers*

```{r table6, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
load(file.path(path.input2, "misplacedMarkers.Rdata"))
out=data.frame(misplacedMarkers)
df=cbind(1:ncol(out),colnames(out))

Class=c()
for(i in 1:ncol(out))Class[i]=class(out[,i])
df=cbind(df,Class)
                   
colnames(df)<-c("Column no.","Column name","Class")                  
 
# output the table in a format suitable for HTML/PDF/docx conversion
datatable(df,rownames=F,options=list(searching=T,dom=""))

```

<font size="2"> 
snp: SNP name <br>
old_chr: Chromosome number according to [ARS-UCD1.2](https://bovinegenome.elsiklab.missouri.edu/downloads/ARS-UCD1.2) <br>
old_bp: Chromosomal position in bp according to [ARS-UCD1.2](https://bovinegenome.elsiklab.missouri.edu/downloads/ARS-UCD1.2)  <br>
new_chr: Chromosome number based on linkage disequilibrium (LD) analysis. Negative values indicate inconclusive results and the analysis continued on the original chromosome. <br>
new_bp: Chromosomal position denotes the coordinate of the SNP to which the minimum recombination rate was achieved <br>
theta: Minimum recombination rate on new chromosome<br>
clear_recrate: Coded as 1 if the recombination rate < 0.01 at the new coordinate (bp), zero otherwise; in total 34 <br>
BLAST_probeA_chr.bp: The position of alternate match when the probe sequence of candidate SNP was aligned to the reference genome. No entry implies a single hit coinciding with the map coordinate in [ARS-UCD1.2](https://bovinegenome.elsiklab.missouri.edu/downloads/ARS-UCD1.2) or no hit. <br>
candidate:  Coded as 1 if the SNP was identified as candidate of misplacement based on estimates of recombination rate in the original study [Qanbari and Wittenburg (2020)](https://doi.org/10.1186/s12711-020-00593-z); in total 51 <br>
problemregion:  Identifier of coherent problematic region; in total 15<br>
start: Start of position of problematic region<br>
end: End of position of problematic region<br>
n_heterozygous: Number of heterozygous sires based on which recombination rates were estimated <br>
</font> 

&nbsp;&nbsp;

### 2.4 generalproblematicRegions.Rdata

tibble *generalProblematicRegions*

```{r table8, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
load(file.path(path.input2, "generalProblematicRegions.Rdata"))
out=data.frame(generalProblematicRegions)
df=cbind(1:ncol(out),colnames(out))

Class=c()
for(i in 1:ncol(out))Class[i]=class(out[,i])
df=cbind(df,Class)
                   
colnames(df)<-c("Column no.","Column name","Class")                  
 
# output the table in a format suitable for HTML/PDF/docx conversion
datatable(df,rownames=F,options=list(searching=F,dom=""))
```
<font size="2"> 
Chr: Chromosome <br>
Start: Start of position of problematic region in bp <br>
End: End of position of problematic region in bp <br>
HD: Number of SNPs on Illumina HD array falling in this region <br>
GGPF250: Number of SNPs on GGP-F250 array falling in this region <br>
SNP50: Number of SNPs on Illumina SNP50 array falling in this region <br>
</font> 
