---
title: "*CLARITY*: A Shiny app for interactive visualisation of the bovine physical-genetic map"
author: "N. Melzer, D. Wittenburg"
date: "(December 18, 2025)"
output: 
 github_document
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = 400
)
```

<!-- badges: start  -->
[![R CMD Check](https://github.com/nmelzer/CLARITY/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/nmelzer/CLARITY/actions/workflows/R-CMD-check.yaml)
⁠<!-- badges: end -->

## Source

An online version of CLARITY is available at https://nmelzer.shinyapps.io/clarity/

A pipeline for creating a breed-specific genetic map from genotypes of half-siblings is available at <https://github.com/wittenburg/hsrecombi> ([v2.0.1](https://doi.org/10.5281/zenodo.17962688))


## Background 
The arrangement of markers on the genome can be defined in either physical or linkage terms. While a physical map represents the inter-marker distances in base pairs, a genetic (or linkage) map pictures the recombination rate between pairs of markers. High-resolution genetic maps are key elements for genomic research, such as fine-mapping of quantitative trait loci, but they are also needed for creating and updating chromosome-level assemblies of whole-genome sequences. 

## Objectives

Frequencies of recombination events have been studied in 8 European
cattle breeds comprising dairy, dual-purpose and beef breeds (German
Holstein, Swiss Holstein, German/Austrian Fleckvieh, Brown Swiss,
Original Braunvieh, Simmental, Limousin, Angus). Here, we provide a
platform that allows users to interactively explore the bovine genetic
and physical map in each breed and to compare between breeds. We developed the R Shiny app “CLARITY”
that provides access to the genetic map built on the Illumina Bovine
SNP50 genotyping array (or similar panel) with markers ordered according
to the physical coordinates of the most recent bovine genome assembly
[ARS-UCD1.2](https://bovinegenome.elsiklab.missouri.edu/downloads/ARS-UCD1.2).
The user is able to interconnect the physical and genetic map for a
whole chromosome or a specific chromosomal region and can inspect a
landscape of recombination hotspots. Comparison of genetic maps estimated from three different approaches is implemented: male, female and sex-average genetic maps from Hidden-Markov model, male genetic maps from likelihood-based approach, and
male genetic maps from deterministic approach. Moreover, the user can investigate
which of the frequently used genetic-map functions locally fits best based on the likelihood-based approach. We
further provide auxiliary information about markers being putatively
misplaced in the
[ARS-UCD1.2](https://bovinegenome.elsiklab.missouri.edu/downloads/ARS-UCD1.2)
release. Genetic maps and corresponding output tables and figures can be downloaded
in various formats. The app facilitates comparison of different
genome features among cattle breeds, providing a valuable tool for
education and research purposes.

Key features:

- Summary statistics for all chromosomes
- Male, female, and sex-average genetic maps
- Hotspot detection
- Genetic-map functions
- Misplaced markers


## Installation

To get the current development version of CLARITY from github, please type:

```{r install}
# install.packages("remotes") (if not installed)
remotes::install_github("nmelzer/CLARITY")
```

## Example

To start the R Shiny app, please type:

```{r example,eval=FALSE}
library(CLARITY)
run_app()
```


## Cite
Code and app (v3.0.0): [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.11620801.svg)](https://doi.org/10.5281/zenodo.11620801) 

Data (v2.0): [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.17909700.svg)](https://doi.org/10.5281/zenodo.17909700)
 


### Note
The data in CLARITY [v2.0.0](https://github.com/nmelzer/CLARITY/tree/v2.0.0) was still in the development phase and should not be used.

